### PR TITLE
feat: init commit, add contracts design, basic tests and config

### PR DESCRIPTION
**Pull request by cosinlink**
1. CKBChiain which contains headers-chain and spv-prover
2. TokenLocker which contains lockToken and unlockToken
3. decode lib on TypedMemView.sol, ViewCKB and ViewSpv
3. tests on ./test and ./scripts
4. solium config on .soliumrc.json
5. buidler config on ./buidler.config.js

### CODE REVIEW

File: eth-contracts/.soliumrc.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      "error"
+    ],
+    "security/no-block-members": [
+      "off"
+    ],
+    "security/no-inline-assembly": [
+      "warning"
+    ]
+  }
+}
```
Comment (huwenchao) [inline eth-contracts/.soliumrc.json:49]:
```suggestion
}

```

File: eth-contracts/contracts/TokenLocker.sol [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    // before lockToken, user should approve -> TokenLocker Contract with 0xffffff token
+    function lockToken(address token, uint256 amount, string memory ckbAddress) public {
+        // TODO modify `transferFrom` to `safeTransferFrom`
+        IERC20(token).transferFrom(msg.sender, address(this), amount);
+        emit Locked(token, msg.sender, amount, ckbAddress);
+    }
+
+    function lockETH(uint256 amount, string memory ckbAddress) public payable {
```
Comment (huwenchao) [inline eth-contracts/contracts/TokenLocker.sol:68]: What is the `amount` for?

File: eth-contracts/contracts/TokenLocker.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    function unlockToken(bytes memory proofData) public {
+        require(ckbSpv_.proveTxExist(proofData, numConfirmations_), "tx from proofData should exist");
+
+        // Unpack the proof and extract the execution outcome.
+        bytes29 proof = proofData.ref(uint40(ViewSpv.SpvTypes.TransactionProof));
+
+        // TODO modify `mockTxHash` to `txHash`
+        bytes32 txHash = proof.mockTxHash();
```
Comment (huwenchao) [inline eth-contracts/contracts/TokenLocker.sol:79]: Why do you use `mockTxHash` instead of `txHash` here?

Comment (cosinlink) [reply eth-contracts/contracts/TokenLocker.sol:79]: proof struct will be changed  after,  this will be fixed today

File: eth-contracts/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# CKBChain
+a ckb light client and spv prover on ethereum
+
+# TokenLocker
+TokenLocker is linked to the fungible token on Ethereum side and mintable fungible
+token on CKB side, it also links to the prover that it uses to unlock the tokens.
+
+## Quick Start
```
Comment (mkxbl) [inline eth-contracts/README.md:8]:
```suggestion
The ethereum contracts component of forece-bridge-eth, mainly contains `CKBChain` and `TokenLocker`.

# CKBChain
CKBChain is a ckb light client and spv prover.

# TokenLocker
TokenLocker locks erc-20 token or eth, and emits events, which can be used to mint sUDT token on CKB. In reverse, it handle bridge message(sUDT burned) from CKB, and unlock erc-20 token or eth if the message is verified.

# Quick Start
```

**Comment by huwenchao** [comment]
The test failed when following quick start.
image

File: eth-contracts/README.md (hunk 4 above)
Comment (huwenchao) [reply eth-contracts/README.md:8]: There is usually only one level 1 title (`#`) in a doc.

File: eth-contracts/contracts/interfaces/ICKBChain.sol [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        uint64 indexed blockNumber,
+        bytes32 blockHash
+    );
+
+    event BlockHashReverted(
+        uint64 indexed blockNumber,
+        bytes32 blockHash
+    );
+
+    function addHeader(bytes calldata data) external;
```
Comment (mkxbl) [inline eth-contracts/contracts/interfaces/ICKBChain.sol:14]: inconsistent with impl method
```suggestion
    function addHeaders(bytes calldata data) external;
```

File: eth-contracts/package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@cosinlink/ckb-spv",
+  "version": "0.0.1",
+  "description": "CKB SPV proofs in Solidity",
+  "scripts": {
+    "compile": "truffle compile",
+    "test": "cp ../testVectors.json test/ && truffle test",
+    "test:coverage": "cp ../testVectors.json test/ && npx solidity-coverage",
+    "lint": "solium -d contracts/ && eslint ./test",
+    "lint:fix": "solium --fix -d contracts/ && eslint --fix ./test"
```
Comment (huwenchao) [inline eth-contracts/package.json:10]: This commands do not work.

**Comment by huwenchao** [comment]
LGTM

**Pull request merged**